### PR TITLE
Retrieve segments concurrently

### PR DESCRIPTION
**Pull request by bigsheeper**
issue: https://github.com/milvus-io/milvus/issues/24244

/kind enhancement

### CODE REVIEW

File: internal/querynodev2/segments/retrieve.go [hunk 1]
```diff
@@ -18,31 +18,71 @@ package segments
...
-		}
-		result, err := segment.Retrieve(ctx, plan)
+	searchLabel := metrics.SealedSegmentLabel
+	if segType == commonpb.SegmentState_Growing {
+		searchLabel = metrics.GrowingSegmentLabel
+	}
+
+	for i, segID := range segIDs {
+		wg.Add(1)
+		go func(segID int64, i int) {
```
Comment (yah01) [inline internal/querynodev2/segments/retrieve.go:49]: Recommend to use `errgroup`, which will stop as saw the first error, also make the code simpler. But LGTM

**Review by yah01 — COMMENTED** [review]
/lgtm

File: internal/querynodev2/segments/retrieve.go (hunk 1 above)
Comment (bigsheeper) [reply internal/querynodev2/segments/retrieve.go:49]: Ok, I'll modify it as well as the errs in search.go in the next pr

File: internal/querynodev2/segments/retrieve.go [hunk 2]
```diff
@@ -18,31 +18,71 @@ package segments
...
+		wg       sync.WaitGroup
+	)
 
-	for _, segID := range segIDs {
-		segment, _ := manager.Segment.Get(segID).(*LocalSegment)
-		if segment == nil {
-			continue
-		}
-		result, err := segment.Retrieve(ctx, plan)
+	searchLabel := metrics.SealedSegmentLabel
```
Comment (czs007) [inline internal/querynodev2/segments/retrieve.go:42]: queryLabel

**Comment by bigsheeper** [comment]
rerun ut

**Comment by xiaofan-luan** [comment]
we probably need to think of adding more metrics in the retrieve path

**Comment by bigsheeper** [comment]
> we probably need to think of adding more metrics in the retrieve path

@xiaofan-luan After this pr, we have the following metrics for the retrieve path:
1. Query latency in the proxy.
2. Reduce query results latency in the proxy.
3. Query latency in each QueryNode.
4. Query latency for a single segment in the QueryNode (inclusive of retrieving the output field from ChunkManager).
5. Latency for a single retrieving CGO call.

**Pull request merged**